test/integration/types_test.rb

```diff
@@ -268,6 +268,38 @@ def test_escaped_data
     end
   end
 
+  def test_copy_data_in_fixed_size_binary_column
+    [source_db, target_db].each do |db|
+      db.query("CREATE DATABASE IF NOT EXISTS #{DEFAULT_DB}")
+      db.query("CREATE TABLE IF NOT EXISTS #{DEFAULT_FULL_TABLE_NAME} (id bigint(20) not null auto_increment, data BINARY(4), primary key(id))")
+    end
+
+    # NOTE: We explicitly test with trailing 0s, because the MySQL replication master will strip
+    # such trailing 0s when streaming events to us. As a result, the binlog writer must explicitly
+    # add then when building update/delete statements, as the WHERE clause would not match existing
+    # rows in the target DB
+    inserted_data = "ABC\x00"
+    updated_data = "EFGH"
+
+    source_db.query("INSERT INTO #{DEFAULT_FULL_TABLE_NAME} (id, data) VALUES (1, _binary'#{inserted_data}')")
+
+    ghostferry = new_ghostferry(MINIMAL_GHOSTFERRY)
+
+    ghostferry.on_status(Ghostferry::Status::ROW_COPY_COMPLETED) do
```

Comment:
Can you use `BINLOG_STREAMING_STARTED` instead? Should be less racy that way.

Reply:
hm, I actually think I tried that. Maybe I misunderstood what's the sequence of events. 

When is binlog streaming started in this scenario - after the copy is complete or before that? I thought the former. Thus, I really want the test to happen _after_ we are copying, because the bug occurs only during binlog streaming, not during the copy (as far as I can tell)

Reply:
Binlog streaming and data copy begin and occur simultaneously.

Reply:
ACK

Reply:
when the `ROW_COPY_COMPLETED` event is sent, the binlog streamer is [about to be terminated](https://github.com/Shopify/ghostferry/blob/9443a28/test/lib/go/integrationferry.go#L138-L145). I don't recall 100%, but I think there's a strong race possibility that the binlog is not picked up by Ghostferry.

When these events are sent and these ruby blocks are called, the Ghostferry code is blocked from proceeding until the ruby block returns. As such, a more fool-proof test would be something like this:

```
source_db.query("INSERT INTO ... (binary data...)")
row_copy_called = false
binlog_appled_called = false
ghostferry.on_status(Ghostferry::Status::AFTER_ROW_COPY) do
  row_copy_called = true
  # select row from the target and then make sure the data with 0 padding is present
  # Perform the update query to set the data to `updated_data`.
end

ghostferry.on_status(Ghostferry::Status::AFTER_BINLOG_APPLY) do
  binlog_apply_called = true
  # select row from the target and make sure the data is on the updated_data
end

# run ghostferry,
# assert table identical
assert binlog_apply_called
assert row_copy_called
# if we want to be defensive, query the target and make sure the data is updated_data.

```

Reply:
I tried your suggestion, but am running into two issues with this:

first, updating the source-DB from the context of `AFTER_ROW_COPY` fails. Do we hold some type of lock at the time this is invoked?

second, when checking in `AFTER_BINLOG_APPLY`, it seems binlog applying isn't fully done yet. Thus, it keep seeing the old data and the validation fails

I understand your proposed code, but seems that's equally racy or something else is off

Reply:
Oh yeah you're right. Sorry for the faulty suggestion. In AFTER_BINLOG_APPLY there's indeed a lock due to SELECT FOR UPDATE. I misremembered when I made that suggestion. For now using `BINLOG_STREAMING_STARTED` should work slightly better. At some point in the future we can make the "breakpoints" a little bit more robust and less racy.

Reply:
got it, thanks for the details. No problem for the misdirection - glad I wasn't doing anything wrong

to avoid a misunderstanding: so is there anything else that we should fix?

Reply:
Other than using `BINLOG_STREAMING_STARTED`, not now.

Reply:
ok, then there seems to have been a misunderstanding - that's why I was asking if streaming was started in parallel with the row-copy above (and you answered they are started in parallel...)

if we use `BINLOG_STREAMING_STARTED` , we race in the unit-test (and actually usually lose the race), meaning that callback is called before the row-copy has inserted the original data in the target.